mtsp/src/main.leo

```diff
@@ -0,0 +1,673 @@
+import credits.aleo;
+
+// The 'mtsp' program.
+program mtsp.aleo {
+  record Token {
+    owner: address,
+    amount: u128,
+    token_id: field,
+    authorized: bool,
+    authorized_until: u32
+  }
+
+  struct TokenMetadata {
+    token_id: field,
+    name: u128, // ASCII text represented in bits, and the u128 value of the bitstring
+    symbol: u128, // ASCII text represented in bits, and the u128 value of the bitstring
+    decimals: u8,
+    supply: u128,
+    max_supply: u128,
+    admin: address,
+    external_authorization_required: bool // whether or not this token requires authorization from an external program before transferring
+  }
+
+  struct TokenOwner {
+    account: address,
+    token_id: field
+  }
+
+  struct Balance {
+    token_id: field,
+    account: address,
+    balance: u128,
+    authorized_until: u32
+  }
+
+  struct Allowance {
+    account: address,
+    spender: address,
+    token_id: field
+  }
+
+  mapping registered_tokens: field => TokenMetadata;
+  mapping balances: field => Balance; // hash(token_id, account) => Balance
+  mapping authorized_balances: field => Balance; // hash(token_id, account) => Balance
+  mapping allowances: field => u128; // hash(token_id, account, spender) => Allowance
+
+  // -------------------------
+  // Called by token admins
+  // -------------------------
+
+  transition register_token(token_id: field, name: u128, symbol: u128, decimals: u8, max_supply: u128, external_authorization_required: bool) {
+    let token: TokenMetadata = TokenMetadata {
+      token_id: token_id,
+      name: name,
+      symbol: symbol,
+      decimals: decimals,
```

Comment:
It would be good to think carefully about decimals limitations. Realistically only tokens with decimals=~6 can work well inside AMM pools because we have multiplications in our formulas and u128 is pretty small.

Reply:
Yes, I'm not sure if it should live here though. You can make a check to lower the limit from 255 to something like 10 but I'm thinking we leave it up to the developers.